File: baml_language/sdk_tests/crates/java/type_shapes/customizable/roundtrip_tests/TestLiterals.java
// Roundtrip coverage for `baml_sdk.literals` — literal Ty variants.
//
// Float literals are intentionally absent (BAML's parser rejects a bare
// negative-literal-as-field, and Java has no distinct float-literal type
// anyway). The negative-literal-as-field case is still parser-blocked, but
// the function-return form `return_literal_neg_one() -> -1` emits and is
// exercised here.
//
// Port of python_pydantic2/type_shapes/customizable/roundtrip_tests/
// test_literals.py — same test names, cases, inputs, assertions.
//
// java-port note: BAML literal types (`42`, `"draft"`, `true`) have no
// runtime representation distinct from their base primitive in Java — Java
// has no literal/refinement type system, unlike Python's `Literal[...]`
// (which pyright still statically narrows). Every literal-typed value below
// is therefore just a plain `long` / `String` / `boolean`.
package roundtrip_tests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import baml_sdk.literals.Fns;
import baml_sdk.literals.Literals;
import org.junit.jupiter.api.Test;

class TestLiterals {

    @Test
    void test_literals_return_literals() {
        assertEquals(42L, Fns.return_literal42());
        assertEquals(-1L, Fns.return_literal_neg_one());
        assertEquals("draft", Fns.return_literal_draft());
        assertEquals("has \"quotes\"", Fns.return_literal_escaped());
        assertTrue(Fns.return_literal_true());
        assertFalse(Fns.return_literal_false());
    }

    @Test
    void test_literals_round_trip_literal42() {
        assertEquals(42L, Fns.round_trip_literal42(42L));
    }

    @Test
    void test_literals_round_trip_literal_draft() {
        assertEquals("draft", Fns.round_trip_literal_draft("draft"));
    }

    @Test
    void test_literals_round_trip_literal_escaped() {
        assertEquals("has \"quotes\"", Fns.round_trip_literal_escaped("has \"quotes\""));
    }

    @Test
    void test_literals_round_trip_literal_true() {
        assertTrue(Fns.round_trip_literal_true(true));
    }

    @Test
    void test_literals_round_trip_literal_false() {
        assertFalse(Fns.round_trip_literal_false(false));
    }

    @Test
    void test_literals_round_trip_literals() {
        Literals lit = new Literals(42L, "draft", "has \"quotes\"", true, false);
        assertEquals(lit, Fns.round_trip_literals(lit));
    }
}

File: baml_language/sdk_tests/crates/rust/type_shapes/customizable/roundtrip_tests/test_literals.rs
//! Roundtrip coverage for `baml_sdk::literals` — literal Ty variants.
//!
//! Float literals are intentionally absent (Python `Literal` rejects
//! floats). The negative-literal-as-field case is still parser-blocked, but
//! the function-return form `return_literal_neg_one() -> -1` emits and is
//! exercised here.

// PROVISIONAL: literal types have no Rust SDK design yet. This port assumes
// they surface as their base primitive types (i64 / String / bool) in
// generated signatures and struct fields.
use baml_sdk::literals::{
    Literals, return_literal_draft, return_literal_escaped, return_literal_false,
    return_literal_neg_one, return_literal_true, return_literal42, round_trip_literal_draft,
    round_trip_literal_escaped, round_trip_literal_false, round_trip_literal_true,
    round_trip_literal42, round_trip_literals,
};

#[test]
fn test_literals_return_literals() {
    assert_eq!(return_literal42().unwrap(), 42);
    assert_eq!(return_literal_neg_one().unwrap(), -1);
    assert_eq!(return_literal_draft().unwrap(), "draft");
    assert_eq!(return_literal_escaped().unwrap(), "has \"quotes\"");
    assert!(return_literal_true().unwrap());
    assert!(!return_literal_false().unwrap());
}

#[test]
fn test_literals_round_trip_literal42() {
    assert_eq!(round_trip_literal42(42).unwrap(), 42);
}

#[test]
fn test_literals_round_trip_literal_draft() {
    assert_eq!(
        round_trip_literal_draft("draft".to_string()).unwrap(),
        "draft"
    );
}

#[test]
fn test_literals_round_trip_literal_escaped() {
    assert_eq!(
        round_trip_literal_escaped("has \"quotes\"".to_string()).unwrap(),
        "has \"quotes\""
    );
}

#[test]
fn test_literals_round_trip_literal_true() {
    assert!(round_trip_literal_true(true).unwrap());
}

#[test]
fn test_literals_round_trip_literal_false() {
    assert!(!round_trip_literal_false(false).unwrap());
}

#[test]
fn test_literals_round_trip_literals() {
    let lit = Literals {
        literal_42: 42,
        literal_draft: "draft".to_string(),
        literal_escaped: "has \"quotes\"".to_string(),
        literal_true: true,
        literal_false: false,
    };
    assert_eq!(round_trip_literals(lit.clone()).unwrap(), lit);
}
